src/llvm-ptls.cpp

```diff
@@ -169,17 +182,29 @@ void LowerPTLS::fix_pgcstack_use(CallInst *pgcstack, Function *pgcstack_getter,
         if (CFGModified)
             *CFGModified = true;
         // emit slow branch code
-        CallInst *adopt = cast<CallInst>(pgcstack->clone());
         Function *adoptFunc = M->getFunction(XSTR(jl_adopt_thread));
         if (adoptFunc == NULL) {
-            adoptFunc = Function::Create(pgcstack_getter->getFunctionType(),
+            adoptFunc = Function::Create(FunctionType::get(builder.getPtrTy(), { builder.getPtrTy()}, false),
                 pgcstack_getter->getLinkage(), pgcstack_getter->getAddressSpace(),
                 XSTR(jl_adopt_thread), M);
             adoptFunc->copyAttributesFrom(pgcstack_getter);
             adoptFunc->copyMetadata(pgcstack_getter, 0);
         }
-        adopt->setCalledFunction(adoptFunc);
-        adopt->insertBefore(slowTerm);
+        builder.SetInsertPoint(slowTerm);
+        Value* handle = Constant::getNullValue(builder.getPtrTy());
+        if (imaging_mode) {
+            // Adopt thread takes in a handle to the sysimage and this is the easiest way to get it.
+            Function *dladdr = M->getFunction(XSTR(jl_find_dynamic_library_by_addr)); // gets handle to sysimage
+            if (dladdr == NULL) {
+                dladdr = Function::Create(FunctionType::get(builder.getPtrTy(), { builder.getPtrTy()}, false),
+                    pgcstack_getter->getLinkage(), pgcstack_getter->getAddressSpace(),
+                    XSTR(jl_find_dynamic_library_by_addr), M);
+            }
+            auto this_func = builder.GetInsertBlock()->getParent();
+            handle = builder.CreateCall(dladdr, {ConstantExpr::getBitCast(this_func, builder.getPtrTy())});
```

Comment:
Do we ever use this in pkgimages?

It seem like it will get the wrong handle in that case.